suite-native/discovery/src/discoveryThunks.ts

```diff
@@ -1,24 +1,247 @@
+import { getDerivationType } from '@suite-common/wallet-utils';
+import { Network } from '@suite-common/wallet-config';
+import { DiscoveryStatus } from '@suite-common/wallet-constants';
 
-export const createAndStartDiscoveryThunk = createThunk(
-    `${DISCOVERY_MODULE_PREFIX}/createAndStart`,
+import { DeviceAccessMutex, getBundleDescriptors } from './utils';
+
+const DISCOVERY_ROUND_SIZE = 3;
+
+type DiscoveryDescriptorItem = DiscoveryItem & { descriptor: string };
+
+const finishNetworkTypeDiscoveryThunk = createThunk(
+    `${DISCOVERY_MODULE_PREFIX}/incrementDiscoveryProgressThunk`,
+    (_, { dispatch, getState }) => {
+        const discovery = selectDiscoveryForDevice(getState());
+
+        if (!discovery) {
+            return;
+        }
+
+        const finishedNetworksCount = discovery.loaded + 1;
+        dispatch({
+            type: updateDiscovery.type,
+            payload: {
+                ...discovery,
+                loaded: finishedNetworksCount,
+            },
+        });
+
+        if (finishedNetworksCount === discovery.networks.length) {
+            dispatch({
+                type: updateDiscovery.type,
+                payload: {
+                    ...discovery,
+                    status: DiscoveryStatus.COMPLETED,
+                },
+            });
+        }
+    },
+);
+
+const discoverAccountsByDescriptorThunk = createThunk(
+    `${DISCOVERY_MODULE_PREFIX}/discoverAccountsByXpubThunk`,
     async (
-        { deviceState, device }: { deviceState: string; device: TrezorDevice },
+        {
+            descriptorsBundle,
+            deviceState,
+        }: {
+            descriptorsBundle: DiscoveryDescriptorItem[];
+            deviceState: string;
+        },
         { dispatch },
     ) => {
+        let isFinalRound = false;
+
+        if (!descriptorsBundle) {
+            isFinalRound = true;
+        }
+
+        // eslint-disable-next-line no-restricted-syntax
+        for (const bundleItem of descriptorsBundle) {
+            // eslint-disable-next-line no-await-in-loop
+            const { success, payload: accountInfo } = await TrezorConnect.getAccountInfo({
+                coin: bundleItem.coin,
+                descriptor: bundleItem.descriptor,
+            });
+
+            if (success) {
+                if (accountInfo.empty) {
+                    isFinalRound = true;
+                    break;
+                }
+
+                dispatch(
+                    accountsActions.createAccount(
+                        deviceState,
+                        bundleItem,
+                        accountInfo,
+                        false,
+                        // TODO: name of newly created account will be handled in follow-up issue:
+                        // see more: https://github.com/trezor/trezor-suite/issues/9501
+                        bundleItem.descriptor,
+                    ),
+                );
+            }
+        }
+
+        return isFinalRound;
```

Comment:
I'd argue to rethink the design of this thunk. It's an antipattern to return boolean values from async thunks most of the time. You're trying to say if getAccount info was succesfull or not, could you just return fulfilled / rejected or https://mobily.github.io/ts-belt/api/result/ ?

Reply:
Won't be changed after a mutual discussion.